src/commands/init/index.ts

```diff
@@ -74,39 +48,85 @@ Velocitas project found!
-            }
-            this.log(`... Downloading package: '${packageConfig.getPackageName()}:${packageConfig.version}'`);
-            await packageConfig.downloadPackageVersion(verbose);
-        }
-    }
-
     async run(): Promise<void> {
         const { flags } = await this.parse(Init);
-
         this.log(`Initializing Velocitas packages ...`);
+        const projectConfig = this.initializeProject();
+
+        const appManifestData = AppManifest.read(projectConfig.getVariableMappings().get(APP_MANIFEST_PATH_VARIABLE));
+
+        await this.ensurePackagesAreDownloaded(projectConfig, flags.force, flags.verbose);
+
+        if (!flags['no-hooks']) {
+            await this.runPostInitHook(projectConfig, appManifestData, flags.verbose);
+        }
+
+        this.createProjectLockFile(projectConfig, flags.verbose);
+    }
+
+    initializeProject(): ProjectConfig {
         let projectConfig: ProjectConfig;
 
         if (!ProjectConfig.isAvailable()) {
             this.log('... Directory is no velocitas project. Creating .velocitas.json at the root of your repository.');
             projectConfig = new ProjectConfig(`v${this.config.version}`);
             projectConfig.write();
+        } else {
+            projectConfig = ProjectConfig.read(`v${this.config.version}`, undefined, true);
         }
-        projectConfig = ProjectConfig.read(`v${this.config.version}`);
+        return projectConfig;
+    }
 
-        const appManifestData = AppManifest.read(projectConfig.getVariableMappings().get(APP_MANIFEST_PATH_VARIABLE));
+    async ensurePackagesAreDownloaded(projectConfig: ProjectConfig, force: boolean, verbose: boolean) {
+        for (const packageConfig of projectConfig.getPackages()) {
+            const packageVersions = await packageConfig.getPackageVersions();
+            const packageVersion = getMatchedVersion(packageVersions, packageConfig.version);
 
-        await this.ensurePackagesAreDownloaded(projectConfig, flags.force, flags.verbose);
+            if (verbose) {
+                this.log(`... Resolved '${packageConfig.getPackageName()}:${packageConfig.version}' to version: '${packageVersion}'`);
+            }
 
-        if (!flags['no-hooks']) {
-            projectConfig.validateUsedComponents();
-            for (const componentContext of projectConfig.getComponents()) {
+            packageConfig.setPackageVersion(packageVersion);
+
+            if (!force && packageConfig.isPackageInstalled()) {
+                this.log(`... '${packageConfig.getPackageName()}:${packageConfig.version}' already installed.`);
+                continue;
+            }
+
+            this.log(`... Downloading package: '${packageConfig.getPackageName()}:${packageConfig.version}'`);
+            await packageConfig.downloadPackageVersion(verbose);
+        }
+    }
+
+    async runPostInitHook(projectConfig: ProjectConfig, appManifest: any, verbose: boolean) {
+        projectConfig.validateUsedComponents();
```

Comment:
Method should have one responsiblity: Running all post init hooks. This validation step should happen before this call or in `initializeOrReadProject`